ledger/metrics/src/main/scala/com/daml/telemetry/TelemetryContext.scala

```diff
@@ -0,0 +1,235 @@
+
+import com.daml.dec.DirectExecutionContext
+import io.grpc.MethodDescriptor
+import io.opentelemetry.api.common.AttributeKey
+import io.opentelemetry.api.trace.Span
+import io.opentelemetry.context.Context
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+trait TelemetryContext {
+
+  /** Sets or replaces the value of attribute `key` to `value`.
+    */
+  def setAttribute[T](key: AttributeKey[T], value: T): TelemetryContext
+
+  /** Creates a new span and runs the computation inside it.
+    * The new span has its parent set as the span associated with the current context.
+    * A new context containing the new span is passed as parameter to the computation.
+    *
+    * @param spanName   the name of the new span
+    * @param kind       the kind of the new span
+    * @param attributes the key-value pairs to be set as attributes to the new span
+    * @param body       the computation to be run in the new span
+    * @return the result of the computation
+    */
+  def runFutureInNewSpan[T](
+      spanName: String,
+      kind: SpanKind,
+      attributes: (AttributeKey[String], String)*
+  )(
+      body: TelemetryContext => Future[T]
+  ): Future[T]
+
+  /** Creates a new span and runs the computation inside it.
+    * The new span has its parent set as the span associated with the current context.
+    * A new context containing the new span is passed as parameter to the computation.
+    *
+    * @param spanName the name of the new span
+    * @param kind the kind of the new span
+    * @param attributes the key-value pairs to be set as attributes to the new span
+    * @param body the computation to be run in the new span
+    * @return the result of the computation
+    */
+  def runInNewSpan[T](
+      spanName: String,
+      kind: SpanKind,
+      attributes: (AttributeKey[String], String)*
+  )(
+      body: TelemetryContext => T
+  ): T
+
+  /** Runs the computation inside an OpenTelemetry scope.
+    *
+    * This is used to set the tracing metadata in the gRPC local thread context, so it
+    * can be accessed and used by the OpenTelemetry instrumentation to create complete traces.
+    *
+    * It should be used around gRPC calls, to ensure that the tracing metadata is correctly used
+    * and transferred.
+    *
+    * @param body the computation to be run in the Telemetry scope
+    * @return the result of the computation
+    */
+  def runInOpenTelemetryScope[T](body: => T): T
+
+  /** Encode the metadata of the context in a key-value map.
+    *
+    * Typically, metadata is encoded in a map to allow transporting it across process boundaries.
+    * Originally, it has been created to carry tracing metadata across boundaries, and
+    * to create complete traces.
+    *
+    * @see [[com.daml.telemetry.Telemetry.contextFromMetadata(java.util.Map)]]
+    */
+  def encodeMetadata(): jMap[String, String]
```

Comment:
For example in [CantonSyncService.submitTransaction](https://github.com/DACH-NY/canton/blob/master/community/participant/src/main/scala/com/digitalasset/canton/participant/sync/CantonSyncService.scala#L208) we create a new [TraceContext](https://github.com/DACH-NY/canton/blob/master/community/common/src/main/scala/com/digitalasset/canton/tracing/TraceContext.scala#L22)
Instead of creating a fresh context, we'd like to create `TraceContext(opentelemetryContextFromWriteService)`

Reply:
@danilofaria I cannot open the link, is there a chance I could get read access to the repo? I'm a bit fresh to this API but I think the idea was to abstract over the opentelemetry lib and not to use bare objects from its API. What you might like to use `Telemetry.runInSpan` or `Telemetry.runFutureInSpan` which are the basic building blocks and allow you to get a fresh context. Or we might need to have a more detailed conversation on how the API should look like.